File: python-algorithm/leetcode/problem_540.py
"""540. Single Element in a Sorted Array
https://leetcode.com/problems/single-element-in-a-sorted-array/
"""
from typing import List


def single_non_duplicate(nums: List[int]) -> int:
    ans = 0
    for num in nums:
        ans ^= num
    return ans


def single_non_duplicate2(nums: List[int]) -> int:
    lo, hi = 0, len(nums) - 1
    while lo < hi:
        mid = (lo + hi) // 2
        both_even = (hi - mid) % 2 == 0
        if nums[mid] == nums[mid + 1]:
            if both_even:
                lo = mid + 2
            else:
                hi = mid - 1
        elif nums[mid] == nums[mid - 1]:
            if both_even:
                hi = mid - 2
            else:
                lo = mid + 1
        else:
            return nums[mid]
    return nums[lo]

File: java-algorithm/src/main/java/com/leetcode/Problem540.java
package com.leetcode;

/**
 * 540. Single Element in a Sorted Array
 * https://leetcode.com/problems/single-element-in-a-sorted-array/
 */
public class Problem540 {
    public int singleNonDuplicate(int[] nums) {
        int i = 0, j = nums.length - 1;
        while (i < j) {
            int mid = i + (j - i) / 2;
            if (mid % 2 == 0) {
                if (mid > 0 && nums[mid] == nums[mid - 1]) {
                    j = mid - 2;
                } else if (mid < nums.length - 1 && nums[mid] == nums[mid + 1]) {
                    i = mid + 2;
                } else {
                    return nums[mid];
                }
            } else {
                if (nums[mid] == nums[mid - 1]) {
                    i = mid + 1;
                } else if (mid < nums.length - 1 && nums[mid] == nums[mid + 1]) {
                    j = mid - 1;
                } else {
                    return nums[mid];
                }
            }
        }
        return nums[i];
    }
}
